### DISCUSSION TITLE
0008 XLS-8d Tickets: Supporting out-of-sequence transactions on the XRP Ledger

### DISCUSSION THREAD

**Discussion by nbougalis** [XRPLF/XRPL-Standards#23] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by WietseWind** [comment]
Really good. This would have my vote in a heartbeat, very useful (especially in Multi Sign scenarios indeed). Thank you!

+1 for `TicketSignerQuorum`

And @ ledger spam: maybe add reserve > x open Tickets?

**Comment by mDuo13** [comment]
There are a lot of annoying exceptions and special cases that tickets add. Fortunately, this spec walks through a lot of them in great detail. 👌

Unfortunately, if this becomes enabled, I'll have to ctrl-f the docs for all the cases I talk about how a transaction is guaranteed to increase its sender's sequence number and add a note about Tickets to each one. 😫

**Comment by timethernom** [comment]
Great to see this moving forward.  Have you thought about representing the tickets with a bitfield?  E.g., add an STBlob to the account root with 1 bit for each contiguous ticket sequence number.  Set the corresponding bit when a ticket is used.  This would mitigate the issue of ledger spam.  N tickets would add only 3 + N/8 bytes to the account root.

It seems like this method could also be used to create tickets on demand.  E.g., if the next expected sequence number, aSeq, is 10, and sequence number 14 arrives, leave aSeq unchanged but add a bitfield to the account root set to 0x08.  If 10 arrives next, shift the bitfield right by 1 and increment aSeq.  Delete the bitfield when it shifts to 0.  This would not be appropriate for all transactions, so could be activated by a new tx flag.

**Comment by RichardAH** [comment]
It's an interesting solution. I feel like it is a very complex solution and a simpler one might exist.
Perhaps I'm missing something but could we change the sequence number semantics inside transactions to include `any of` [**X**, **Y**, **Z**] instead of `only` **X**?

Suppose Bob and Alice both wish to submit a transaction at roughly the same time on behalf of the same shared account. They both go to the account state and get the next sequence number **X** and specify it in the transaction. But, being aware that a second signer exists, they additionally specify that **X+1** could be used instead of **X**. Thus the transaction's field reads `Sequence: any of [X, X+1]`

To prevent replay attacks, first check if all the sequence numbers specified in the transaction have already been used. If they have then no replay is possible. If some haven't been used, then check the hashes of the transactions on the sequence numbers that have been used. If any of these hashes matches the current transaction then it is being replayed and must be dropped. If none of the hashes match the transaction then it is accepted and it takes the lowest available sequence number in its `any of` list.

For simplicity and to avoid a bunch of other issues the `any of` list should be a consecutive range starting with a recent sequence number. Additionally, to ensure the validators still have the recent transactions and ledgers needed to perform the checks above the transaction must include `LastLedgerSequence` no larger than some specified value that ensures these will be available. As a final backstop against abuse a node will refuse to apply any transaction whose `any of` sequence numbers can't be checked for replay attacks.

**Comment by nbougalis** [comment]
Hey @codetsunami. Thanks for the interesting suggestion.

>To prevent replay attacks, first check if all the sequence numbers specified in the transaction have already been used. If they have then no replay is possible. If some haven't been used, then check the hashes of the transactions on the sequence numbers that have been used. If any of these hashes matches the current transaction then it is being replayed and must be dropped. If none of the hashes match the transaction then it is accepted and it takes the lowest available sequence number in its any of list.

It's certainly a neat way of tackling the problem, and simpler than tickets in some sense but I think that its simplicity limits it enough that it fails to address the more interesting use-cases tickets is meant to solve. Imagine, for example, a corporate pre-signing several hundred transactions to be executed in the future (not the _near_ future, mind; perhaps in a few years) in an order yet to be determined. This solution would prevent said entity from using that account for anything else, until the pre-signed transactions all cleared.

**Comment by RichardAH** [comment]
A second alternative to tickets is to introduce a way to share capital between XRPL accounts:

TX sequence number conflicts can be resolved by having a separate account for each **line of** transactions that might be issued. So an organisation that needs to sign say 200 transactions in advance can do so on Account `A` while using Account `B` as an ordinary day to day transaction account. Account `A` contains most of the organisation's funds and Account `B` is linked to `A` through a special trustline that allows XRP to be drawn down up to a specified limit, on demand, as required to satisfy a given transaction entering a ledger.

Aside from alleviating the need for tickets this would also introduce some defi properties to the XRPL including potentially lending. It would also provide an additional mechanism for securing funds in operational accounts. I.e. by setting the limit on the special trustline as to how much can be drawn down for the day/month and then by updating it the next day/month, potentially even with a pre-signed transaction that is released by a custody bot.

This solution has multiple benefits and provides far less code and far less attack surface area than ticketing.

**Comment by intelliot** [comment]
This spec has been replaced by [XLS-13d](https://github.com/XRPLF/XRPL-Standards/discussions/18)

**Comment by mvadari** [comment]
@nbougalis should this spec be moved into PR and kept in the repo or should it be closed? Note that this does not mean that the spec has been accepted/implemented/finalized.

I will move it to PR by copy-pasting the existing draft if there is no response in the next week.
